### PR TITLE
Increase coverage

### PR DESCRIPTION
**Pull request by szmarczak**

### CODE REVIEW

**Comment by szmarczak** [comment]
I'll leave this part:

https://github.com/sindresorhus/got/blob/6ba9e68c7f496a69641a8aac51be0781f1ae0eac/source/timed-out.js#L30-L32

this makes us sure `timed-out` isn't called twice or more, although it's obvious we don't do it.

**Comment by szmarczak** [comment]
We also need tests for `electron` support.

**Comment by szmarczak** [comment]
I just wonder if this piece of code is needed:

https://github.com/sindresorhus/got/blob/6ba9e68c7f496a69641a8aac51be0781f1ae0eac/source/get-body-size.js#L31-L33

because I tried passing a `Buffer` as body and this doesn't even get called.

File: source/normalize-arguments.js [hunk 1]
```diff
@@ -165,10 +165,10 @@ module.exports = (url, options, defaults) => {
 
 			if (Reflect.has(error, 'headers') && Reflect.has(error.headers, 'retry-after') && retryAfterStatusCodes.has(error.statusCode)) {
 				let after = Number(error.headers['retry-after']);
-				if (is.number(after)) {
-					after *= 1000;
+				if (is.nan(after)) {
+					after = Date.parse(error.headers['retry-after']) - Date.now();
```
Comment (szmarczak) [inline source/normalize-arguments.js]: This fixes a bug: at first I thought `is.number` won't return `true` for NaN, but I was wrong.

`Math.max` is not needed BTW.

File: source/request-as-event-emitter.js [hunk 2]
```diff
@@ -142,13 +142,7 @@ module.exports = (options = {}) => {
 
 		if (backoff) {
 			retryCount++;
-			setTimeout(options => {
-				try {
-					get(options);
-				} catch (error2) {
-					emitter.emit('error', error2);
-				}
-			}, backoff, options);
```
Comment (sindresorhus) [inline source/request-as-event-emitter.js]: Why was this removed?

**Comment by sindresorhus** [comment]
> because I tried passing a Buffer as body and this doesn't even get called.

You have to pass a stream. The buffer testing is testing a property on the stream.

File: source/request-as-event-emitter.js (hunk 2 above)
Comment (szmarczak) [reply source/request-as-event-emitter.js]: `get()` does not throw any errors directly. It uses `emitter.emit('error', error);`

**Comment by szmarczak** [comment]
@sindresorhus But it's converted to a stream here:

https://github.com/sindresorhus/got/blob/6ba9e68c7f496a69641a8aac51be0781f1ae0eac/source/normalize-arguments.js#L111-L114

File: source/request-as-event-emitter.js (hunk 2 above)
Comment (sindresorhus) [reply source/request-as-event-emitter.js]: Good point. I think we need a try/catch here though: https://github.com/sindresorhus/got/pull/542/files#diff-62bdc57f6f22ae58f495daef16f21f8bR141 In case the `options.gotRetry.retries` function throws.

**Comment by sindresorhus** [comment]
@szmarczak Hmm, then maybe it returns here: https://github.com/sindresorhus/got/blob/6ba9e68c7f496a69641a8aac51be0781f1ae0eac/source/get-body-size.js#L10 Add some `debugger` statements to each `if` and figure out where it returns.

File: source/request-as-event-emitter.js (hunk 2 above)
Comment (sindresorhus) [reply source/request-as-event-emitter.js]: Your answer is not confusing, but you're not answering my question. Too be clear, I think we need to try/catch this line because it could throw: `const backoff = options.gotRetry.retries(++retryTries, error);`

**Comment by szmarczak** [comment]
> Hmm, then maybe it returns here:

Yeah, you're right, checked it. So, is it needed? IMO we can get rid of that.

File: source/request-as-event-emitter.js (hunk 2 above)
Comment (szmarczak) [reply source/request-as-event-emitter.js]: > you're not answering my question.

Sorry, if you're talking about the second one then it hadn't popped up when I was writing the answer. Pardon my confusion.

> I think we need to try/catch this line because it could throw

Good point! Indeed. 🙌

Comment (sindresorhus) [reply source/request-as-event-emitter.js]: > I think we need to try/catch this line because it could throw

Comment (sindresorhus) [reply source/request-as-event-emitter.js]: >> I think we need to try/catch this line because it could throw
> Good point! Indeed. 🙌

And of course a test ;)

Comment (szmarczak) [reply source/request-as-event-emitter.js]: Done! I hope it satisfies you :)

**Comment by sindresorhus** [comment]
> Yeah, you're right, checked it. So, is it needed? IMO we can get rid of that.

See: https://github.com/sindresorhus/got/blob/4d92eb6bef37637c0d85986faae314e17e9cd9bc/source/request-as-event-emitter.js#L163-L166

File: test/retry.js [hunk 3]
```diff
@@ -247,3 +249,13 @@ test('works when defaults.options.retry is not an object', async t => {
...
 });
+
+test('retry function can throw', async t => {
+	await t.throws(got(`${s.url}/413`, {
+		retry: {
+			retries: () => {
+				throw new Error('Simple error');
+			}
+		}
+	}), 'Simple error');
```
Comment (sindresorhus) [inline test/retry.js]: It's a good practice to put the message in a variable and use it in both to make absolutely sure they are the same and stay the same in the future too.

Comment (szmarczak) [reply test/retry.js]: Yeah.

**Comment by szmarczak** [comment]
> This should be done in a separate PR though

So it will be :)

> It's a good practice to put the message in a variable and use it in both to make absolutely sure they are the same and stay the same in the future too.

Done :unicorn:

**Comment by sindresorhus** [comment]
> We also need tests for electron support.

Nah, don't care. If it prevents us from getting 100% test coverage, we can just mock it.

> Note: Missing test for #490 and for:

Do you intend to do that in this PR? Asking so I know whether this is ready to merge or not.

**Comment by sindresorhus** [comment]
Something is failing here: https://ci.appveyor.com/project/sindresorhus/got/build/1.0.245/job/nunr63w0rttwjv0t

**Comment by szmarczak** [comment]
>  If it prevents us from getting 100% test coverage

The point is: not only that keeps us from 100%. But it's possible to achieve 99% :)

> Do you intend to do that in this PR? Asking so I know whether this is ready to merge or not.

No, that'll be a separate PR too. Yes, it's ready to merge now :)

**Comment by szmarczak** [comment]
See? Fixed :)

**Comment by sindresorhus** [comment]
Yup, AppVeyor is annoying. We had so many problems with it over at https://github.com/avajs/ava too...

**Comment by sindresorhus** [comment]
👍 I'm extra happy that doing this actually caught a couple of potential bugs.
